Cassandra 12937 sstable compression rebased to  trunk

Claudenw opened the pull request:
Pull request 2199 rebased on trunk.

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

Associated pull request: https://github.com/apache/cassandra/pull/2199

Claudenw committed:
initial code

Claudenw committed:
first cut

Claudenw committed:
Working implementation with tests

Claudenw committed:
cleaned up

Claudenw committed:
fixed name typo

Claudenw committed:
removed accidental formatting

Claudenw committed:
removed accidental formatting

Claudenw committed:
Update conf/cassandra.yaml

Co-authored-by: mck <mck@apache.org>

Claudenw committed:
Changes to use xKiB style yaml entries

Claudenw committed:
cleaned up/similified code

Claudenw committed:
fixed formatting

Claudenw committed:
cleaned up formatting

Claudenw committed:
moved sstable_compressor section

Claudenw committed:
updates as per review

Claudenw committed:
cleaned up formatting

Claudenw committed:
fixed sstable compression default to match earlier default

Claudenw committed:
cleaned up merge errors

Claudenw committed:
cleanedup code

Claudenw committed:
updated to remove CompressionParams from config structures

Claudenw committed:
fixed typo

Claudenw committed:
created CompressionParamsTest

Claudenw commented:
@smiklosovic Please take a look at using this as the basis for the final patch.

Claudenw committed:
cleaned up CompressionParams processing

Claudenw committed:
updated to use calculated values

Claudenw committed:
fixed tests

Claudenw committed:
removed unused import

Claudenw committed:
fixed comment

Claudenw committed:
fixed comment

Claudenw committed:
cleaned up code

Claudenw commented:
@smiklosovic Please review this pull request.  I think it is complete and correct.

michaelsembwever commented on conf/cassandra.yaml:
```diff
@@ -566,6 +566,35 @@ commitlog_segment_size: 32MiB
+# length and min_compress_ratio of 0.0 (see CompressionParams.DEFAULT_MIN_COMPRESS_RATIO).
+# To disable compression set `type: none` or set the `enable` parameter as `false`.
+# This is the default options for sstable compression, the (schema) table-level settings will override this option.
+#sstable_compressor:
+#
+  #enable=false # defaults to true
+#
+  #chunk_length: 16KiB : defaults to 16KiB
+#
+  # min_compress_ratio: 0.0
+  # max_compressed_length: 16KiB
```
For readability, make the example standalone, add comments afterwards.
```suggestion
#
# sstable_compressor:
  # enable=true 
  # chunk_length: 16KiB
  # min_compress_ratio: 0.0
  # max_compressed_length: 16KiB
  # - class_name:
  # parameters:
     # - param1 : value
#     
# 
# enable defaults to true 
# chunk_length defaults to 16KiB
# Only one of the min_compress_ratio and max_compressed_length options can be specified.  
# The mathematically related in that
# chunk_length / max_compressed_length = min_compress_ratio
# if neither option is specified a min_compress_ratio of 0.0 is the default.
# A min_compress_ratio <= 0 implies max_compressed_length of Integer.MAX_VALUE KiB
# The class_name is the compressor class name. It may be one of the following aliases:
# lz4,  none,  noop,  snappy,  deflate,  or zstd; or it the fully qualified  name of a class
# that implements ICompressor
```

Claudenw committed:
removed SSTableCompressionOptions class

Claudenw committed:
working implementation

Claudenw committed:
added round trip map test

Claudenw committed:
updated configuration documentation

michaelsembwever commented on conf/cassandra.yaml:
```diff
@@ -566,6 +566,48 @@ commitlog_segment_size: 32MiB
+#         max_comrpessed_length: 16KiB
+#         class_specific_parameter: value
+#
+#
+# The class_name is the compressor class name. It may be one of the aliases,
+# the class name of a system ICompressor implementation, or fully qualified
+# name of a class that implements ICompressor.
+#
+# class aliases are:
+#  alias   system compressor impl.
+#  lz4      LZ4Compressor,
+#  none     (null) -- disabled
+#  noop     NoopCompressor
+#  snappy   SnappyCompressor
+#  deflate  DeflateCompressor
+#  zstd     ZstdCompressor
+#
+# The standard parameters are any required or optional parameter for the instantiation of the
+# specified class, or one of the following standard parameters:
+# parameter                       usage
+# enable                Disables compression if set to false. Defaults to true.
+# chunk_length          The lenght of the compresson chunks, defaults to 16KiB
+# min_compress_ratio    The miniimal acceptable compression, must greater than or equal to 1.0.
+# max_compressed_length The maximum size for a compressed block.  Must be less than
+#                           chunk_length.   Defaults to Integer.MAX_VALUE
+# chunk_length_in_kb    The CQL compression parameter.
+#
+# Only one of the min_compress_ratio and max_compressed_length options can be specified.
+# They are  mathematically related in that
+# chunk_length / max_compressed_length = min_compress_ratio
+# if neither option is specified a min_compress_ratio of 0.0 is the default.
+#
+# Only one of chunk_length or chunk_length_in_kb may be specified.
```
nit: my bad, please put the comment section before the commented out config. this is the norm in this file.

Claudenw replied:
Fixed

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -566,6 +566,48 @@ commitlog_segment_size: 32MiB
 # flush_compression: fast
 
+#
+sstable_compressor:
```
@Claudenw this should be `sstable_compression`. We have `commitlog_compression` as well as `hints_compression`. We have even `internode_compression` and `flush_compression`. No reason to have `sstable_compressor` instead of `sstable_compression`.

Claudenw replied:
fixed

smiklosovic commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -1087,6 +1087,9 @@ public static PaxosStatePurging fromBoolean(boolean enabled)
     public volatile boolean top_partitions_enabled = true;
 
+    @Nullable
+    public ParameterizedClass sstable_compressor;
```
sstable_compressor -> sstable_compression

Claudenw replied:
fixed

smiklosovic commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -191,6 +191,9 @@
 
     private static Map<String, Supplier<SSTableFormat<?, ?>>> sstableFormatFactories;
 
+    /** The sstable compression options */
```
The usefulness of this comment is questionable. It is pretty obvious this is about SSTable compression from the name of that variable itself.

Claudenw replied:
removed

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
-
-        return cp;
+        static CompressorType forClass(String name) {
+            for (CompressorType type : CompressorType.values()) {
```
We could quickly return null if `name` is null so we are not looping unnecessarily. Please keep the braces on the new line in this whole method as well.

Claudenw replied:
fixed.  Also reorganized the Enum order so that `none` is last since we are capturing it first.

smiklosovic commented on test/conf/cassandra-with-sstable-compressor.yaml:
```diff
@@ -0,0 +1,110 @@
+        test_invalid_factory_field:
+            class_name: org.apache.cassandra.cql3.validation.operations.CreateTest$InvalidMemtableFactoryField
+
+sstable_compressor:
```
Please, dont forget to rename this as well to `sstable_compression`.

Claudenw replied:
fixed

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -55,35 +58,44 @@
+    @VisibleForTesting
+    static final String TOO_MANY_CHUNK_LENGTH = "Only one of 'chunk_length', 'chunk_length_kb', or 'chunk_length_in_kb' may be specified";
+
     @Deprecated public static final String SSTABLE_COMPRESSION = "sstable_compression";
```
@Claudenw I would remove all these deprecated options as part of this PR. At least `crc_check_chance` seems to be deprecated in 3.0 so it is eligible for deletion. I think that the rest is good to remove too but we would need to confirm this as git history is little bit unclear here.

cc @michaelsembwever

Claudenw replied:
I removed all the deprecated options.  However crc_check_chance is set and read by functions that are used elsewhere so the variable itself can not be removed.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
-            sstableCompressionClass = null;
+    public enum CompressorType
+    {
+        lz4("LZ4Compressor"),
```
@Claudenw what if names of these classes are changed or we make mistake typing them out? Would not it be better if we use `LZ4Compressor.class.getName()` instead?

Claudenw replied:
done

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
-        else
-        {
-            sstableCompressionClass = removeSstableCompressionClass(options);
+        String className;
```
this might be final

Claudenw replied:
done

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
-        {
-            sstableCompressionClass = removeSstableCompressionClass(options);
+        String className;
+        CompressorType(String className) {
```
please take care of braces on new lines in this whole enum.

Claudenw replied:
I believe that I have fixed the braces and space changes requested through out this review.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
-        return new CompressionParams((ICompressor) null, DEFAULT_CHUNK_LENGTH, Integer.MAX_VALUE, 0.0, Collections.emptyMap());
-    }
+        if (options == null)
+        {
```
braces around one return statement are not necessary, are they?

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
 
-    // The shorthand methods below are only used for tests. They are a little inconsistent in their choice of
-    // parameters -- this is done on purpose to test out various compression parameter combinations.
+        if (options.parameters != null && options.parameters.containsKey(SSTABLE_COMPRESSION))
```
this should probably go away if we remove all deprecated stuff

Claudenw replied:
All deprecated stuff removed.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
-        return new CompressionParams(LZ4Compressor.create(Collections.emptyMap()), chunkLength, maxCompressedLength, calcMinCompressRatio(chunkLength, maxCompressedLength), Collections.emptyMap());
-    }
+        if (!removeEnabled(options))
+        {
```
I do not understand why it is not done just like this:

````
        boolean enabled = removeEnabled(options);
````

Claudenw replied:
holdover from old code.  Fixed now.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -270,6 +339,10 @@ public int chunkLength()
         return chunkLength;
     }
 
+    double minCompressRatio() {
```
nit: braces on new line

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+        // calculate the max_compressed_length and min_compress_ratio
+        int max_compressed_length;
+        double min_compress_ratio;
+        String max_compressed_length_str = options.remove( MAX_COMPRESSED_LENGTH);
```
please remove these spaces before / after argument in a method. There are various instances of that in this method in general.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+                validateMaxCompressedLength( max_compressed_length, chunk_length_in_kb);
+                min_compress_ratio = CompressionParams.calcMinCompressRatio(chunk_length_in_kb, max_compressed_length);
+            } catch (IllegalArgumentException e) {
+                throw new ConfigurationException(invalidValue( MAX_COMPRESSED_LENGTH, e.getMessage(), max_compressed_length_str ));
```
here too etc ... more cases like this.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+        }
+
+        // try to set compressor type
+        CompressorType compressorType = StringUtils.isEmpty(sstableCompressionClass)?CompressorType.lz4:null;
```
`StringUtils.isEmpty(sstableCompressionClass) ? CompressorType.lz4 : null;`

Here some spaces would be handy!

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -402,29 +475,55 @@ private static Integer parseChunkLength(String chLengthKB) throws ConfigurationE
      */
     private static int removeChunkLength(Map<String, String> options)
```
this whole method could use some love on formatting, spaces, braces, no braces if one statement only ...

Claudenw replied:
I believe that I have fixed the braces and space changes requested through out this review.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+            switch (compressorType)
+            {
+                case none:
+                    cp = new CompressionParams((ICompressor) null, chunk_length_in_kb, max_compressed_length, min_compress_ratio, options);
```
IDEA tells me that you do not need to cast `null` to `ICompressor`. It can be just `null`.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+                    cp = new CompressionParams(enabled?LZ4Compressor.create(options):null, chunk_length_in_kb, max_compressed_length, min_compress_ratio, options);
+                    break;
+                case snappy:
+                    cp = new CompressionParams(enabled?SnappyCompressor.instance:null, chunk_length_in_kb, max_compressed_length, min_compress_ratio, options);
```
shouldnt be here `SnappyCompressor.create(options)` instead of `SnappyCompressor.instance`? Same for DeflateCompressor.

Claudenw replied:
The test factory does it this way, but I have reverted to create method.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+            }
+        } else {
+            cp = new CompressionParams(enabled?sstableCompressionClass:null, options, chunk_length_in_kb, min_compress_ratio);
+            if (enabled && cp.getSstableCompressor() == null)
```
This might go away when deprecated stuff goes away as well.

smiklosovic commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -97,128 +109,185 @@
+        if (compressorType != null)
+        {
+
+            switch (compressorType)
```
I wonder why we need this entire `switch` at all. If we treat all compressors as equal, why do we need to have a switch and then a special `else` branch? Why not to instantiate all compressors the same way as it is done for case of a custom one? I just do not like that we are making the difference between in-built compressors and custom ones. A compressor is just a compressor. The fact that it is accidentally in-built is irrelevant. 

It will be also aligned with how we are doing it in cassandra.yaml, conceptually. If we are making a flat map and we treat all compressors equal, irrelevant where they are from, we would just apply the same reasoning there and it would simplify the code as well.

Claudenw replied:
Reworked to remove switch and clean up the process.

Claudenw committed:
changes for review

Claudenw committed:
fixed tests

Claudenw committed:
cleaned up config and code

Claudenw committed:
fixed typo and cleaned up code

michaelsembwever commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -402,29 +411,47 @@ private static Integer parseChunkLength(String chLengthKB) throws ConfigurationE
-            if (options.containsKey(CHUNK_LENGTH_KB))
+        Integer chunk_length_in_kb = null;
+        String key = null;
+        if (options.containsKey(CHUNK_LENGTH)) {
```
nit: brace on newline.

michaelsembwever commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -402,29 +411,47 @@ private static Integer parseChunkLength(String chLengthKB) throws ConfigurationE
-                                                        CHUNK_LENGTH_KB,
-                                                        CHUNK_LENGTH_IN_KB));
+                chunk_length_in_kb =  new DataStorageSpec.IntKibibytesBound(value).toKibibytes();
+            } catch (IllegalArgumentException e) {
```
nit: braces on newline.

michaelsembwever commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -402,29 +411,47 @@ private static Integer parseChunkLength(String chLengthKB) throws ConfigurationE
+        }
 
-            return parseChunkLength(options.remove(CHUNK_LENGTH_KB));
+        if (chunk_length_in_kb != null) {
```
nit: brace on newline.

michaelsembwever commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -518,23 +527,36 @@ private static boolean removeEnabled(Map<String, String> options)
-    public void validate() throws ConfigurationException
-    {
-        // if chunk length was not set (chunkLength == null), this is fine, default will be used
+    private static void validateChunkLength(String key, int chunkLength) throws ConfigurationException{
```
nit: brace on newline.

michaelsembwever commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -518,23 +527,36 @@ private static boolean removeEnabled(Map<String, String> options)
+            throw new ConfigurationException(invalidValue(key,  "Must be a power of 2", chunkLength));
+    }
 
+    private static void validateMinCompressRatio(double ratio) throws ConfigurationException {
```
nit: brace on newline.

michaelsembwever commented on src/java/org/apache/cassandra/schema/CompressionParams.java:
```diff
@@ -518,23 +527,36 @@ private static boolean removeEnabled(Map<String, String> options)
+
+    }
+
+    private static void validateMaxCompressedLength(int maxCompressedLength, int chunkLength) throws ConfigurationException {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/io/sstable/SSTableReaderTest.java:
```diff
@@ -33,6 +33,10 @@
 import com.google.common.collect.Sets;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.io.util.File;
```
nit: wrong location

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    //private CompressionParams params;
+
+
+    private static  ParameterizedClass emptyParameterizedClass() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+
+    @Test
+    public void additionalParamsTest() {
+        // no map
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+
+
+    @Test
+    public void constructorTest() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+                                                               .withMessageContaining("Invalid 'max_compressed_length' value for the 'compression' option.  May not be less than zero: -1");
+    }
+
+    private static void roundTripMapTest(CompressionParams params) {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    @Test
+    public void lz4Test() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    @Test
+    public void noneTest() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    @Test
+    public void noopTest() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    @Test
+    public void snappyTest() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    @Test
+    public void deflateTest() {
```
nit: brace on newline.

michaelsembwever commented on test/unit/org/apache/cassandra/schema/CompressionParamsTest.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    @Test
+    public void zstdTest() {
```
nit: brace on newline.

Claudenw committed:
There was a path where unrecognised variables were not detected.  That has been corrected and tests added.

There was an issue with chunk_length_in_kb not being serialized to the map correctly in some cases.  That has been corrected and tests added.

The commented out configuration in the yaml file is not functional if uncommented.

Additional comments were added to explain the items that were previously explained by the broken configuration. (e.g. max_compressed_length needing to have the GiB, MiB, KiB suffix)

Additional comment was added to specify that all parameter values must be strings.

Claudenw committed:
removed unused logging

Claudenw committed:
fixed includes and curlies

Claudenw committed:
rebased to head

Claudenw commented:
No longer current.  Changes merged into pull #3168
